core/src/sumeragi/fault.rs

```diff
@@ -294,8 +262,133 @@ impl<F: FaultInjection> SumeragiWithFault<F> {
         }
         panic!("Timed out waiting for peers to come online");
     }
+
+    pub fn pipeline_time(&self) -> Duration {
+        self.block_time + self.commit_time
+    }
+}
+
+fn block_commit<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    block: VersionedValidBlock,
+    state_machine_guard: &mut SumeragiStateMachineData,
+) where
+    F: FaultInjection,
+{
+    let block = block.commit();
+    let block_hash = block.hash();
+
+    if let Err(error) = state_machine_guard.wsv.apply(block.clone()) {
+        error!(%error);
+        panic!("Failed to apply block on WSV. This is absolutely not acceptable.");
+    }
+
+    // Update WSV copy that is public facing
+    {
+        let mut wsv_for_public_use_guard = sumeragi.wsv_for_public_use.lock().unwrap();
+        *wsv_for_public_use_guard = WorldStateView::clone(&state_machine_guard.wsv);
+    }
+
+    for event in Vec::<Event>::from(&block) {
+        trace!(?event);
+        drop(sumeragi.events_sender.send(event));
+    }
+
+    state_machine_guard.latest_block_height = block.header().height;
+    state_machine_guard.latest_block_hash = block.hash();
+
+    // Push new block height information to block_sync
+    *sumeragi
+        .latest_block_hash_for_use_by_block_sync
+        .lock()
+        .expect("lock on latest_block_hash_for_use_by_block_sync") =
+        state_machine_guard.latest_block_hash;
+
+    let previous_role = state_machine_guard.current_topology.role(&sumeragi.peer_id);
+    state_machine_guard
+        .current_topology
+        .refresh_at_new_block(block_hash);
+    info!(
+        prev_peer_role = ?previous_role,
+        new_peer_role = ?state_machine_guard.current_topology.role(&sumeragi.peer_id),
+        new_block_height = %state_machine_guard.latest_block_height,
+        %block_hash,
+        "Committing block"
+    );
+    sumeragi.kura.blocking_store_block(block);
+    SumeragiWithFault::<F>::update_network_topology(
+        &mut state_machine_guard.current_topology,
+        &state_machine_guard.wsv,
+    );
+
+    // Transaction Cache
+    {
+        let mut transaction_cache = &mut state_machine_guard.transaction_cache;
+        // We prune transactions now because now is the only time they may have been
+        // invalidated by a new block being applied to the world state view.
+        //
+        // We also check for if the transaction has expired because we might as well.
+        let mut read_index = 0;
+        let mut write_index = 0;
```

Comment:
These mutable counters are becoming an anti-pattern. But in this _particular_ case the simple functional rewrite actually makes the thing less efficient.

Reply:
Yes. The others have mutable counters because I start off by doing the most flexible thing because I'm not exactly sure what it is I'm making. (This probably means I should do a second pass on my code more often.) In this instance it was a very particular choice in order to be fast.